client/rpcserver/rpcserver.go

```diff
@@ -1,20 +1,405 @@
+}
+
+// Config holds variables neede to create a new RPC Server.
+type Config struct {
+	Core                        ClientCore
+	Addr, User, Pass, Cert, Key string
+}
+
+// SetLogger sets the logger for the RPCServer package.
+func SetLogger(logger slog.Logger) {
 	log = logger
-	log.Infof("RPC server running at %s", addr)
-	<-ctx.Done()
+}
+
+// New is the constructor for an RPCServer.
+func New(cfg *Config) (*RPCServer, error) {
+
+	// Find or create the key pair.
+	keyExists := fileExists(cfg.Key)
+	certExists := fileExists(cfg.Cert)
+	if certExists == !keyExists {
+		return nil, fmt.Errorf("missing cert pair file")
+	}
+	if !keyExists && !certExists {
+		err := genCertPair(cfg.Cert, cfg.Key)
+		if err != nil {
+			return nil, err
+		}
+	}
+	keypair, err := tls.LoadX509KeyPair(cfg.Cert, cfg.Key)
+	if err != nil {
+		return nil, err
+	}
+
+	// Prepare the TLS configuration.
+	tlsConfig := tls.Config{
+		Certificates: []tls.Certificate{keypair},
+		MinVersion:   tls.VersionTLS12,
+	}
+
+	// Create listener.
+	listener, err := tls.Listen("tcp", cfg.Addr, &tlsConfig)
+	if err != nil {
+		return nil, fmt.Errorf("Can't listen on %s. web server quitting: %v", cfg.Addr, err)
+	}
+
+	// Create an HTTP router.
+	mux := chi.NewRouter()
+	httpServer := &http.Server{
+		Handler:      mux,
+		ReadTimeout:  rpcTimeoutSeconds * time.Second, // slow requests should not hold connections opened
+		WriteTimeout: rpcTimeoutSeconds * time.Second, // hung responses must die
+	}
+
+	// Make the server.
+	s := &RPCServer{
+		core:     cfg.Core,
+		listener: listener,
+		srv:      httpServer,
+		syncers:  make(map[string]*marketSyncer),
+		clients:  make(map[int32]*wsClient),
+	}
+
+	// Create authsha to verify requests against.
+	if cfg.User != "" && cfg.Pass != "" {
+		login := cfg.User + ":" + cfg.Pass
+		auth := "Basic " +
+			base64.StdEncoding.EncodeToString([]byte(login))
+		s.authsha = sha256.Sum256([]byte(auth))
+	}
+
+	// Middleware
+	mux.Use(middleware.Recoverer)
+	mux.Use(middleware.RealIP)
+	mux.Use(s.authMiddleware)
+
+	// Websocket endpoint
+	mux.Get("/ws", s.handleWS)
+
+	// https endpoint
+	mux.Post("/", s.handleJSON)
+
+	return s, nil
+}
+
+// Run starts the web server. Satisfies the dex.Runner interface. ctx
+// passed to newMarketSyncer when making new market syncers.
+func (s *RPCServer) Run(ctx context.Context) {
+	s.ctx = ctx
+	// Close the listener on context cancellation.
+	go func() {
+		<-ctx.Done()
+
+		if err := s.srv.Shutdown(context.Background()); err != nil {
+			// Error from closing listeners:
+			err = fmt.Errorf("HTTP server Shutdown: %v", err)
+			fmt.Fprintln(os.Stderr, err)
+		}
+	}()
+	log.Infof("RPC server listening on %s", s.listener.Addr())
+	if err := s.srv.Serve(s.listener); err != http.ErrServerClosed {
+		log.Warnf("unexpected (http.Server).Serve error: %v", err)
+	}
+	s.mtx.Lock()
+	for _, cl := range s.clients {
+		cl.Disconnect()
+	}
+	s.mtx.Unlock()
+
+	// Wait for market syncers to finish.
```

Comment:
Should really also wait for the goroutine with Shutdown in it.  Although Shutdown is what lets Server complete, there is nothing to ensure that the error check and print statements get executed before Run returns (and possible process termination).